proposals/0161-sbpf-feature-gating.md

```diff
@@ -0,0 +1,138 @@
+created: 2024-07-15
+---
+
+## Summary
+
+An explicit versioning system for SBPF programs, that allows enabling or
+disabling features on a per-program basis.
+
+## Motivation
+
+SBPF has evolved over the years and will continue to evolve. Changes are
+necessary among other reasons to improve performance, address security concerns
+and provide new, better features to dapp developers.
+
+Today, the only way to introduce changes to the program runtime is via feature
+gates [link that explains what feature gates are]. For example we have used
+feature gates in the past to restrict new deployments of programs owned by
+deprecated loaders. Feature gates alone are not sufficient to evolve SBPF
+though - we are missing a mechanism to enable (or disable) specific features on
+a per-program basis.
+
+As an example, over two years ago we decided to change the way the stack works
+for SBPF programs - from allocating a fixed 4kB per function frame, to letting
+functions dynamically specify how much stack they need. This change was
+implemented in both the program runtime and the toolchain (LLVM), but as of
+today it has not yet been deployed, because it's essentially too hard to do so:
+upon executing a program, should the program runtime use the old or the new
+stack allocation strategy?
+
+We propose to introduce an explicit versioning scheme for SBPF programs:
+programs will encode which SBPF version they are compiled for; based on this
+version, the program runtime will alter its behavior. In order to prevent an
+"extension hell" it will not be possible to opt into specific changes via
+flags, instead an entire set of changes is reduced to a single SBPF version.
+
+## New Terminology
+
+SBPF version: the SBPF version number a program was compiled for.
+
+## Detailed Design
+
+Every program must signal which SBPF version it was compiled for via
+the `e_flags` field in the
+[ELF header](https://refspecs.linuxfoundation.org/elf/gabi4+/ch4.eheader.html).
+Block explorers are recommended to display this SBPF version field for program
+accounts.
+
+For each SBPF version two feature gates will control its availability:
+
+- One to enable deployment and execution of that SBPF version, which includes
+disabling deployment (but not execution) of all earlier versions.
+- Another to first disincentivize (via CU costs) and later disable execution
+of that SBPF version, so that the runtime and virtual machine can reduce their
+complexity again. This feature gate can be shared by multiple SBPF versions,
+effectively deprecating larger blocks of versions in order to reduce the amount
+of redeployment required from dapp developers.
+
+### Example
+
+| SBPF version | becomes deployable and executable | ceases to be deployable |
+| ------------ | --------------------------------- | ----------------------- |
+| v1           |                                   | Feature gate A          |
+| v2           | Feature gate A                    | Feature gate B          |
```

Comment:
Is this purely an example, or are we considering only one version to be able to deploy at a time? It might be helpful to include the number of simultaneous SBPF versions that the network will allow.

Reply:
Yes the proposal right now says that there can only be one deployable version at a time (but many executable versions). See the discussion above.